### PR TITLE
feat(precompile): use reflect to validate precompile implementation arg and return types

### PR DESCRIPTION
**Pull request by shampoobera**
this makes overloaded functions trivial.

### CODE REVIEW

File: eth/core/precompile/method.go [hunk 1]
```diff
@@ -82,12 +82,12 @@ func (m *Method) Call(si StatefulImpl, ctx context.Context, input []byte) ([]byt
...
 		return nil, err
 	}
-	// Build argument list
+	// convert the any unnamed struct types into `execute`'s requested struct type
 	reflectedUnpackedArgs := make([]reflect.Value, 0, len(unpackedArgs))
 	for _, unpacked := range unpackedArgs {
-		reflectedUnpackedArgs = append(reflectedUnpackedArgs, reflect.ValueOf(unpacked))
+		unpackedVal := reflect.ValueOf(unpacked) // geth type
+		reflectedUnpackedArgs = append(reflectedUnpackedArgs, unpackedVal)
 	}
```
Comment (calbera) [inline eth/core/precompile/method.go:90]: dont need any of these changes

File: eth/core/precompile/utils/reflection_validation.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	// last parameter of Go precompile implementation is an error (for reverts),
+	// so we skip that.
+	if implMethod.Type.NumOut()-1 != len(abiMethod.Outputs) {
+		return errors.New("number of return types mismatch")
+	}
+
+	// if the function does not take any inputs, no need to check.
+	// note again that for NumIn(), we check for 2 args, because the first two are the receiver and
+	// ctx due to the nature of Go's `reflect` package.
+	if len(abiMethod.Inputs) == 0 && implMethod.Type.NumIn() == 2 {
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:63]: redundant check, the second condition will always be true if the first condition is true. bc of line 50

File: eth/core/precompile/utils/reflection_validation.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+				return err
+			}
+		} else if implMethodVarType.Field(j).Type != abiMethodVarType.Field(j).Type {
+			return fmt.Errorf("return type mismatch: %v != %v",
+				implMethodVarType.Field(j).Type,
+				abiMethodVarType.Field(j).Type,
+			)
+		} else {
+			continue
+		}
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:155]: don't need this

File: eth/core/precompile/utils/reflection_validation.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	return nil
+}
+
+// helper function for validateArgumentAndReturnTypes. this function function uses reflection to see
+// what types your implementation uses, and checks against the geth representation of the abi types.
+func validate(implMethodVarType reflect.Type, abiMethodVarType reflect.Type) error {
+	//nolint:exhaustive // nah, this is fine.
+	switch implMethodVarType.Kind() {
+	// if the Go type is any, we leave it to the user to make sure that it is used
+	// correctly.
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:95]: move this comment inside the case

File: eth/core/precompile/utils/reflection_validation.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	// ctx due to the nature of Go's `reflect` package.
+	if len(abiMethod.Inputs) == 0 && implMethod.Type.NumIn() == 2 {
+		return nil
+	}
+
+	j := 0
+	// receiver is 0th param, context is 1st param, so skip those.
+	for i := 2; i < implMethod.Type.NumIn(); i++ {
+		implMethodParamType := implMethod.Type.In(i)
+		abiMethodParamType := abiMethod.Inputs[j].Type.GetType()
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:71]: use `i - 2`

File: eth/core/precompile/factories.go [hunk 6]
```diff
@@ -26,6 +26,7 @@ import (
 
 	"github.com/ethereum/go-ethereum/accounts/abi"
 
+	precompileutils "pkg.berachain.dev/polaris/eth/core/precompile/utils"
```
Comment (calbera) [inline eth/core/precompile/factories.go:29]:
```suggestion
	pcutils "pkg.berachain.dev/polaris/eth/core/precompile/utils"
```

File: eth/core/precompile/utils/reflection_validation.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+			return fmt.Errorf("return type mismatch: %v != %v", implMethodReturnType, abiMethodReturnType)
+		}
+	}
+
+	return nil
+}
+
+// helper function for validateArgumentAndReturnTypes. this function function uses reflection to see
+// what types your implementation uses, and checks against the geth representation of the abi types.
+func validate(implMethodVarType reflect.Type, abiMethodVarType reflect.Type) error {
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:91]: validateArg

File: eth/core/precompile/utils/reflection_validation.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	"fmt"
+	"reflect"
+
+	"pkg.berachain.dev/polaris/eth/accounts/abi"
+)
+
+// validateArgumentAndReturnTypes checks if the precompile method argument and return types match
+// the abi's argument return types. this is for overloaded Solidity functions and general
+// validation.
+func ValidateArgumentAndReturnTypes(implMethod reflect.Method, abiMethod abi.Method) error {
```
Comment (calbera) [inline eth/core/precompile/utils/reflection_validation.go:39]: ValidateMethod()

File: eth/core/precompile/method.go [hunk 9]
```diff
@@ -48,6 +49,8 @@ import (
 // Method is a struct that contains the required information for the EVM to execute a stateful
 // precompiled contract method.
 type Method struct {
+	rcvr StatefulImpl
```
Comment (calbera) [inline eth/core/precompile/method.go:52]: add a comment on what rcvr is and why its needed

File: eth/core/precompile/method_validation.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+package precompile
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// validateArgumentAndReturnTypes checks if the precompile method argument and return types match
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:34]:
```suggestion
// ValidateBasic checks if the precompile method argument and return types match
```

File: eth/core/precompile/method_validation.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+		abiMethodReturnType := abiMethod.Outputs[i].Type.GetType()
+		if err := validateArg(implMethodReturnType, abiMethodReturnType); err != nil {
+			return fmt.Errorf("return type mismatch: %v != %v", implMethodReturnType, abiMethodReturnType)
+		}
+	}
+
+	return nil
+}
+
+// helper function for validateArgumentAndReturnTypes. this function function uses reflection to see
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:88]:
```suggestion
// helper function for ValidateBasic. this function function uses reflection to see
```

File: eth/core/precompile/factories.go [hunk 12]
```diff
@@ -114,22 +114,25 @@ func (sf *StatefulFactory) Build(
 // This function matches each Go implementation of the precompile to the ABI's respective function.
 // It searches for the ABI function in the Go precompile contract and performs basic validation on
 // the implemented function.
-func buildIdsToMethods(pcABI map[string]abi.Method, contractImpl reflect.Value) (map[string]*Method, error) {
+func buildIdsToMethods(
+	si StatefulImpl,
+	pcABI map[string]abi.Method,
```
Comment (calbera) [inline eth/core/precompile/factories.go:119]: remove this and just add `pcABI := si.ABIMethods()` in the function

File: eth/core/precompile/method_validation.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// ValidateBasic checks if the precompile method argument and return types match
+// the abi's argument return types. this is for overloaded Solidity functions and general
+// validation.
+func (m *Method) ValidateBasic() error {
```
Comment (itsdevbear) [inline eth/core/precompile/method_validation.go:37]: @shampoobera great job, this function is excellently commented.

File: eth/core/precompile/method_validation.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+	// receiver is 0th param, context is 1st param, so skip those.
+	for i := 2; i < implMethod.Type.NumIn(); i++ {
+		implMethodParamType := implMethod.Type.In(i)
+		abiMethodParamType := abiMethod.Inputs[i-2].Type.GetType()
+		if err := validateArg(implMethodParamType, abiMethodParamType); err != nil {
+			return fmt.Errorf("argument type mismatch: %v != %v", implMethodParamType, abiMethodParamType)
+		}
+	}
+
+	for i := 0; i < implMethod.Type.NumOut()-1; i++ {
```
Comment (itsdevbear) [inline eth/core/precompile/method_validation.go:77]: add a comment here, what does this for loop do?

File: eth/core/precompile/method.go [hunk 15]
```diff
@@ -48,6 +48,10 @@ import (
 // Method is a struct that contains the required information for the EVM to execute a stateful
 // precompiled contract method.
 type Method struct {
+	// Rcvr is the receiver object of the precompile's executable. This is the object that
+	// implements the respective precompile method.
```
Comment (calbera) [inline eth/core/precompile/method.go:52]:
```suggestion
	// rcvr is the receiver of the method's executable. This is the stateful precompile
	// that implements the respective precompile method.
```

File: eth/core/precompile/method_validation.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+		panic("The Solidity compiler requires all precompile functions to return at least one value. Consider returning a boolean.")
+	}
+
+	// first two args of Go precompile implementation are the receiver contract and the
+	// context, so we skip those.
+	if implMethod.Type.NumIn()-2 != len(abiMethod.Inputs) {
+		return errors.New("number of arguments mismatch")
+	}
+
+	// last parameter of Go precompile implementation is an error (for reverts),
```
Comment (itsdevbear) [inline eth/core/precompile/method_validation.go:55]: This is a huge nit, and something that we should discuss more broadly with @calbera as well. But generally I lean towards proper capitalization and sentence structure for comments. 

I’m good either way, but we decide and then be consistent throughout the repo. 

1) Either proper sentence structure and capitalization (my preference) “SENTENCES”

2) start all comments with lowercase and no periods. “NOT SENTENCES”

File: eth/core/precompile/method.go [hunk 17]
```diff
@@ -75,23 +80,24 @@ func NewMethod(
 
 // Call executes the precompile's executable with the given context and input arguments.
 //
```
Comment (calbera) [inline eth/core/precompile/method.go:82]:
```suggestion
```

File: eth/core/precompile/method.go [hunk 18]
```diff
@@ -59,39 +63,40 @@ type Method struct {
...
 		abiSig:    abiSig,
 		execute:   execute,
 	}
 }
 
 // Call executes the precompile's executable with the given context and input arguments.
-//
-//nolint:revive // needed for reflection.
-func (m *Method) Call(si StatefulImpl, ctx context.Context, input []byte) ([]byte, error) {
+
```
Comment (calbera) [inline eth/core/precompile/method.go:82]:
```suggestion
```

File: eth/core/precompile/method_validation.go (hunk 16 above)
Comment (shampoobera) [reply eth/core/precompile/method_validation.go]: does the linter check for this? think sentences are much cleaner and make the repo look nicer

File: eth/core/precompile/stateful_test.go [hunk 19]
```diff
@@ -55,7 +55,7 @@ var _ = Describe("Stateful Container", func() {
 		ctx = vm.NewPolarContext(
 			context.Background(),
 			vmmock.NewEVM(),
-			common.Address{},
+			common.Address{0},
```
[the remark this answers is not in the document]
Comment (shampoobera) [reply eth/core/precompile/stateful_test.go]: i don't, i was getting Zero value reflection error stuff and thought that was why but forgot to change back

File: eth/core/precompile/method_validation.go (hunk 16 above)
Comment (calbera) [reply eth/core/precompile/method_validation.go]: I would rather avoid over-commenting and only commenting to explain some logic/tradeoff that isnt explicitly made clear by the code. hence most of the comments before a function explain those decisions and effects. Inside the function it should be just enough comments to allow the reader to follow the logic and sentence comments feel a lot "heavier"

File: eth/core/precompile/method_validation.go [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	// If the function does not take any inputs, no need to check.
+	// Note again that for NumIn(), we check for 2 args, because the first two are the receiver and
+	// Context due to the nature of Go's `reflect` package.
+	if implMethod.Type.NumIn() == 2 && len(abiMethod.Inputs) == 0 {
+		return nil
+	}
+
+	// Ceceiver is 0th param, context is 1st param, so skip those.
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:68]:
```suggestion
	// Receiver is 0th param, context is 1st param, so skip those.
```

File: eth/core/precompile/method_validation.go [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+	// Error is the last param, so skip that.
+	// Validate that the precompile return types == abi return types.
+	for i := 0; i < implMethod.Type.NumOut()-1; i++ {
+		implMethodReturnType := implMethod.Type.Out(i)
+		abiMethodReturnType := abiMethod.Outputs[i].Type.GetType()
+		if err := validateArg(implMethodReturnType, abiMethodReturnType); err != nil {
+			return fmt.Errorf("return type mismatch: %v != %v", implMethodReturnType, abiMethodReturnType)
+		}
+	}
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:86]: add the check for the last implMethod's Type Out is of type error

File: eth/core/precompile/method_validation.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+
+	// Validate that our implementation returns an error as the last param.
+	if implMethod.Type.Out(implMethodNumOut-1) != reflect.TypeOf((*error)(nil)).Elem() {
+		return fmt.Errorf("last return type must be error, got %v", implMethod.Type.Out(implMethodNumOut-1))
+	}
+
+	// If the function does not take any inputs, no need to check.
+	// Note again that for NumIn(), we check for 2 args, because the first two are the receiver and
+	// Context due to the nature of Go's `reflect` package.
+	if implMethodNumIn == 2 && abiMethodNumIn == 0 {
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:74]:
```suggestion
	if abiMethodNumIn == 0 {
```

File: eth/core/precompile/method_validation.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+		// The Solidity compiler requires that precompiles must return at least one value.
+		// See https://github.com/berachain/polaris/issues/491 for more information.
+
+		//nolint:lll // error message.
+		panic("The Solidity compiler requires all precompile functions to return at least one value. Consider returning a boolean.")
+	}
+
+	// First two args of Go precompile implementation are the receiver contract and the
+	// Context, so we skip those.
+	if implMethodNumIn-2 != len(abiMethod.Inputs) {
```
Comment (calbera) [inline eth/core/precompile/method_validation.go:56]:
```suggestion
	if implMethodNumIn-2 != abiMethodNumIn {
```

**Comment by itsdevbear** [comment]
this is looking really good, close?

**Review by calbera — APPROVED** [review]
utACK. can we get method_validation unit test cov up? its very specific logic so want to make sure all edge cases are covered.

image

File: eth/core/precompile/factories.go [hunk 24]
```diff
@@ -114,22 +112,28 @@ func (sf *StatefulFactory) Build(
 // This function matches each Go implementation of the precompile to the ABI's respective function.
 // It searches for the ABI function in the Go precompile contract and performs basic validation on
 // the implemented function.
-func buildIdsToMethods(pcABI map[string]abi.Method, contractImpl reflect.Value) (map[string]*Method, error) {
+func buildIdsToMethods(
+	si StatefulImpl,
+	contractImpl reflect.Value,
```
Comment (calbera) [inline eth/core/precompile/factories.go:117]:
```suggestion
	si StatefulImpl, contractImpl reflect.Value,
```

**Comment by shampoobera** [comment]
im gonna add overloaded function support + tests and then merge all together, should be g2g EOD

**Comment by itsdevbear** [comment]
🔥

File: eth/core/precompile/method_helpers.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+// this function finds the ABI method that matches the given impl method. It returns the key in the
+// ABI methods map that matches the impl method.
+func findMatchingABIMethod(
+	implMethod reflect.Method, precompileABI map[string]abi.Method,
+) (string, error) {
+	implMethodName := formatName(implMethod.Name)
+
+	for i := len(implMethodName) - 1; i >= 1; i-- {
```
Comment (itsdevbear) [inline eth/core/precompile/method_helpers.go:38]: is this gonna end us up with some n^3 horribleness?

Comment (shampoobera) [reply eth/core/precompile/method_helpers.go:38]: it's ``O(m*n^2)`` where `m` is the length of the string name and `n` is the # of methods. `m` is usually very small since overloading is rare and `n` is the number of methods. we can assume # of inputs and # of outputs to be constant since they're small.

**Review by calbera — APPROVED** [review]

File: eth/core/precompile/method_test.go [hunk 26]
```diff
@@ -18,7 +18,7 @@
 // MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
 // TITLE.
 
-package precompile_test
+package precompile
```
Comment (itsdevbear) [inline eth/core/precompile/method_test.go:21]: @shampoobera can you break this file up into `method_test.go` && `method_validation_test.go`

File: eth/core/precompile/method_validation.go [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// validateOutputs checks if the impl method output types match the ABI's return types.
+func validateOutputs(implMethod reflect.Method, abiMethod *abi.Method) error {
+	implMethodNumOut := implMethod.Type.NumOut()
+
+	// The Solidity compiler requires that precompiles must return at least one value.
+	// See https://github.com/berachain/polaris/issues/491 for more information.
+	if len(abiMethod.Outputs) == 0 {
+		//nolint:lll // error message.
```
Comment (itsdevbear) [inline eth/core/precompile/method_validation.go:130]: why `lll` is needed?

File: eth/core/precompile/stateful_conatiner_test.go [hunk 28]
```diff
@@ -112,35 +109,9 @@ var _ = Describe("Stateful Container", func() {
 			Expect(err.Error()).To(Equal(
 				"execution reverted: vm error [err during precompile execution] occurred during precompile execution of [getOutputPartial]", //nolint:lll // test.
```
Comment (itsdevbear) [inline eth/core/precompile/stateful_conatiner_test.go:110]: why is lll needed?

**Review by itsdevbear — COMMENTED** [review]
nits

File: eth/core/precompile/method_validation.go (hunk 27 above)
Comment (shampoobera) [reply eth/core/precompile/method_validation.go:130]: the error msg 143 chars long

File: eth/core/precompile/stateful_conatiner_test.go (hunk 28 above)
Comment (shampoobera) [reply eth/core/precompile/stateful_conatiner_test.go:110]: line is 128 characters

File: eth/core/precompile/method_validation.go (hunk 27 above)
Comment (itsdevbear) [reply eth/core/precompile/method_validation.go:130]: can't do on two lines?

File: eth/core/precompile/stateful_conatiner_test.go (hunk 28 above)
Comment (itsdevbear) [reply eth/core/precompile/stateful_conatiner_test.go:110]: can't do on two lines?

File: eth/core/precompile/method_test.go (hunk 26 above)
Comment (shampoobera) [reply eth/core/precompile/method_test.go:21]: image

**Review by itsdevbear — APPROVED** [review]
lgtm!

File: eth/core/precompile/method_validation_test.go [hunk 29]
```diff
@@ -195,5 +206,5 @@ func (m *mockImpl) NumReturnMismatch(_ *big.Int) error {
 }
 
 func (m *mockImpl) ReturnTypeMismatch(_ *big.Int) (string, error) {
-	return "ur mom", nil
```
Comment (itsdevbear) [inline eth/core/precompile/method_validation_test.go]: kek
